File: CPP/Problem_109.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    vector<int> all, finish;

    for (int v = 1; v <= 20; v++) {
        all.push_back(v);       // S
        all.push_back(2*v);    // D
        all.push_back(3*v);    // T
        finish.push_back(2*v); // D only
    }

    all.push_back(25);   // outer bull
    all.push_back(50);   // inner bull (D25)
    finish.push_back(50);

    sort(all.begin(), all.end());

    int ans = 0;

    for (int f : finish) {
        if (f < 100) ans++;

        for (int i = 0; i < (int)all.size(); i++) {
            int a = all[i];
            if (a + f < 100) ans++;
            else continue;

            for (int j = i; j < (int)all.size(); j++) {
                int b = all[j];
                if (a + b + f < 100) ans++;
            }
        }
    }

    cout << ans << "\n"; // 38182
    return 0;
}

File: Java/Problem_109.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        ArrayList<Integer> all = new ArrayList<>();
        ArrayList<Integer> finish = new ArrayList<>();

        for (int v = 1; v <= 20; v++) {
            all.add(v);         // S
            all.add(2*v);      // D
            all.add(3*v);      // T
            finish.add(2*v);   // D only
        }

        all.add(25);   // outer bull
        all.add(50);   // inner bull
        finish.add(50);

        Collections.sort(all);

        int ans = 0;

        for (int f : finish) {
            if (f < 100) ans++;

            for (int i = 0; i < all.size(); i++) {
                int a = all.get(i);
                if (a + f < 100) ans++;
                else continue;

                for (int j = i; j < all.size(); j++) {
                    int b = all.get(j);
                    if (a + b + f < 100) ans++;
                }
            }
        }

        System.out.println(ans); // 38182
    }
}
